core/store/src/trie/mem/updating.rs

```diff
@@ -0,0 +1,1281 @@
+                            let inner_child = UpdatedMemTrieNode::Extension {
+                                extension: existing_key
+                                    .mid(1)
+                                    .encoded(false)
+                                    .into_vec()
+                                    .into_boxed_slice(),
+                                child: old_child,
+                            };
+                            OldOrUpdatedNodeId::Updated(self.new_updated_node(inner_child))
+                        };
+
+                        let mut children = [None; 16];
+                        children[idx as usize] = Some(child);
+                        let branch_node = UpdatedMemTrieNode::Branch { children, value: None };
+                        self.place_node(node_id, branch_node);
+                        // Start over from the same position.
+                        continue;
+                    } else if common_prefix == existing_key.len() {
+                        // Dereference child and descend into it.
+                        let child = self.ensure_updated(old_child);
+                        let node = UpdatedMemTrieNode::Extension {
+                            extension,
+                            child: OldOrUpdatedNodeId::Updated(child),
+                        };
+                        self.place_node(node_id, node);
+                        node_id = child;
+                        partial = partial.mid(common_prefix);
+                        continue;
+                    } else {
+                        // Partially shared prefix. Convert to shorter extension and descend into it.
+                        // On the next step, branch will be created.
+                        let inner_child_node_id =
+                            self.new_updated_node(UpdatedMemTrieNode::Extension {
+                                extension: existing_key
+                                    .mid(common_prefix)
+                                    .encoded(false)
+                                    .into_vec()
+                                    .into_boxed_slice(),
+                                child: old_child,
+                            });
+                        let child_node = UpdatedMemTrieNode::Extension {
+                            extension: existing_key
+                                .encoded_leftmost(common_prefix, false)
+                                .into_vec()
+                                .into_boxed_slice(),
+                            child: OldOrUpdatedNodeId::Updated(inner_child_node_id),
+                        };
+                        self.place_node(node_id, child_node);
+                        node_id = inner_child_node_id;
+                        partial = partial.mid(common_prefix);
+                        continue;
+                    }
+                }
+            }
+        }
+    }
+
+    pub fn delete(&mut self, key: &[u8]) {
```

Comment:
Please add a comment here, mentioning that
* the logic is to go down to the node, if it exists, and then squash nodes on the way up
* key may not exist in trie and it is valid behaviour

Reply:
btw deleting a non-existent key does happen in production, I believe it was for TrieKey type 3, some delayed receipt stuff. It seems that we just issue a delete even though the key may not exist.